### PR TITLE
Patch Series to add loadbalancer TF resources

### PR DESCRIPTION
**Pull request by bpradipt**

### CODE REVIEW

**Comment by bpradipt** [comment]
cc @clnperez @mkumatag ptal.

**Comment by clnperez** [comment]
Two requests: squash down into one commit when ready to merge, and use the same naming convention as suggested here: https://github.com/openshift-powervs/installer/pull/2#pullrequestreview-667369765

Also I think we can remove vars for the worker (since that's created by the control-plane), which I believe you're already looking into.

@mkumatag is the machine-api also responsible for adding the workers to the lb? IOW -- do we need _any_ TF for that at this stage in the install?

**Comment by bpradipt** [comment]
> Two requests: squash down into one commit when ready to merge, and use the same naming convention as suggested here: [#2 (review)](https://github.com/openshift-powervs/installer/pull/2#pullrequestreview-667369765)
> 

I think having separate commits for updating the bootstrap module, adding the loadbalancer module and making changes to main will make it easier to understand the changes going forward instead of one single commit. Do you recommend different PRs for each commit ? I can send separate PRs.

> Also I think we can remove vars for the worker (since that's created by the control-plane), which I believe you're already looking into.

Yeah will remove it

> 
> @mkumatag is the machine-api also responsible for adding the workers to the lb? IOW -- do we need _any_ TF for that at this stage in the install?

I think so. The LB pool needs to be updated with worker node IPs.  Need to look into that.

**Comment by clnperez** [comment]
> I think having separate commits for updating the bootstrap module, adding the loadbalancer module and making changes to main will make it easier to understand the changes going forward instead of one single commit. Do you recommend different PRs for each commit ? I can send separate PRs.

This is true. Maybe in future better to keep different items separated by PR (easier to find in the git log), but since this is already submitted as one we can leave it.

**Comment by bpradipt** [comment]
> > I think having separate commits for updating the bootstrap module, adding the loadbalancer module and making changes to main will make it easier to understand the changes going forward instead of one single commit. Do you recommend different PRs for each commit ? I can send separate PRs.
> 
> This is true. Maybe in future better to keep different items separated by PR (easier to find in the git log), but since this is already submitted as one we can leave it.

I'll edit the title to indicate it's a patch series.

**Comment by bpradipt** [comment]
@clnperez @mkumatag addressed the comments.

File: data/data/powervs/loadbalancer/alb.tf [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+# Licensed Materials - Property of IBM
+#
+# ©Copyright IBM Corp. 2021
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
```
Comment (mkumatag) [inline data/data/powervs/loadbalancer/alb.tf]: I don't see the licence and copyright set for all the other files in the installer repo, not sure if it is intentional?!

File: data/data/powervs/loadbalancer/alb.tf [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+################################################################
+
+locals {
+  api_servers       = concat([var.bootstrap_ip], var.master_ips)
+  api_servers_count = 4 # bootstrap + master 
+  app_servers       = var.master_ips
+  app_servers_count = 3
```
Comment (mkumatag) [inline data/data/powervs/loadbalancer/alb.tf]: I see these counts are hardcoded, I'm wondering whether we can just derive them from the count of var.bootstrap_ip, var.master_ips

File: data/data/powervs/loadbalancer/alb.tf [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+# machine config listener and backend pool
+resource "ibm_is_lb_listener" "machine_config_listener" {
+  lb           = ibm_is_lb.load_balancer.id
+  port         = 22623
+  protocol     = "tcp"
+  default_pool = ibm_is_lb_pool.machine_config_pool.id
+}
+resource "ibm_is_lb_pool" "machine_config_pool" {
+  depends_on = [ibm_is_lb.load_balancer]
```
Comment (mkumatag) [inline data/data/powervs/loadbalancer/alb.tf:27]: do we need this? because `lb = ibm_is_lb.load_balancer.id` will take care the dependecy right?

Comment (mkumatag) [reply data/data/powervs/loadbalancer/alb.tf:27]: same comment applies for all the depends_on statements

File: data/data/powervs/loadbalancer/alb.tf [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+#
+################################################################
+
+locals {
+  api_servers       = concat([var.bootstrap_ip], var.master_ips)
+  api_servers_count = 4 # bootstrap + master 
+  app_servers       = var.master_ips
+  app_servers_count = 3
+}
+
```
Comment (mkumatag) [inline data/data/powervs/loadbalancer/alb.tf:7]: And also we need to have one private load balancer for the internal communication(from workers => masters) for 6443 port

File: data/data/powervs/loadbalancer/versions.tf [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+################################################################
+
+terraform {
+  required_providers {
+    ibm = {
+      source  = "ibm-cloud/ibm"
+      version = "1.24.0"
+    }
+  }
+}
```
Comment (mkumatag) [inline data/data/powervs/loadbalancer/versions.tf]: Can we have one global provider instead of having it in every folder?

File: data/data/powervs/variables.tf [hunk 6]
```diff
@@ -54,38 +54,51 @@ variable "network_name" {
...
+  validation {
+    condition     = length(var.cluster_id) <= 14
+    error_message = "The cluster_id value shouldn't be greater than 14 characters."
+  }
+}
+
+variable "powervs_vpc_name" {
+  type        = string
+  description = "Name of the IBM Cloud Virtual Private Cloud (VPC) to setup the load balancer."
+  default     = "powervs-vpc"
```
Comment (mkumatag) [inline data/data/powervs/variables.tf]: I feel this default is not needed as this is a mandatory param

File: data/data/powervs/variables.tf [hunk 7]
```diff
@@ -54,38 +54,51 @@ variable "network_name" {
...
 }
+
+variable "powervs_vpc_subnet_name" {
+  type        = string
+  description = "Name of the VPC subnet having DirectLink access to the PowerVS private network"
+  default     = "powervs-vpc-subnet"
```
Comment (mkumatag) [inline data/data/powervs/variables.tf]: I feel this default is not needed as this is a mandatory param

**Review by mkumatag — CHANGES_REQUESTED** [review]

File: data/data/powervs/loadbalancer/alb.tf (hunk 2 above)
Comment (bpradipt) [reply data/data/powervs/loadbalancer/alb.tf]: Yeah we can derive from the vars. Wondering if it'll ever change ?

File: data/data/powervs/variables.tf (hunk 6 above)
Comment (bpradipt) [reply data/data/powervs/variables.tf]: What about the case when we want to create the VPC and not use existing VPC?

File: data/data/powervs/loadbalancer/alb.tf (hunk 1 above)
Comment (bpradipt) [reply data/data/powervs/loadbalancer/alb.tf]: I added this for the new files. I think we will need it for all other files.

File: data/data/powervs/loadbalancer/alb.tf (hunk 3 above)
Comment (bpradipt) [reply data/data/powervs/loadbalancer/alb.tf:27]: Right, I think some of the explicit dependencies can be removed.

File: data/data/powervs/loadbalancer/versions.tf (hunk 5 above)
Comment (bpradipt) [reply data/data/powervs/loadbalancer/versions.tf]: Yes

File: data/data/powervs/loadbalancer/alb.tf (hunk 3 above)
Comment (bpradipt) [reply data/data/powervs/loadbalancer/alb.tf:27]: The members are added one by one to the pool and if the explicit dependency is removed the code will get stuck.
So when one member is added the LB status is updating.  If we don't wait for the state to be ready before adding another member there are issues. Using the load balancer `id` doesn't guarantee this behaviour
 Ref: https://registry.terraform.io/providers/IBM-Cloud/ibm/latest/docs/resources/is_lb_listener

In short the resources needs to be added serially.

File: data/data/powervs/loadbalancer/alb.tf (hunk 2 above)
Comment (mkumatag) [reply data/data/powervs/loadbalancer/alb.tf]: default value if 3 but there is scope of changing this value via `controlPlane.replicas` param

File: data/data/powervs/loadbalancer/alb.tf (hunk 3 above)
Comment (mkumatag) [reply data/data/powervs/loadbalancer/alb.tf:27]: Got it, that is the reason even in the UI it allows only one operation at a time, seems like the resource itself can handle one ops at a time.!

File: data/data/powervs/variables.tf (hunk 6 above)
Comment (mkumatag) [reply data/data/powervs/variables.tf]: Then the value should be autogenerated from the cluster-id instead otherwise we won't able to create multiple vpcs with same name in the account.

File: data/data/powervs/loadbalancer/alb.tf [hunk 8]
```diff
@@ -0,0 +1,151 @@
+################################################################
```
Comment (mkumatag) [inline data/data/powervs/loadbalancer/alb.tf]: And one more generic comment for all the code, please add the `cluster-id` tag to all the resources for tearing down purposes.

**Review by mkumatag — CHANGES_REQUESTED** [review]

File: data/data/powervs/loadbalancer/versions.tf (hunk 5 above)
Comment (bpradipt) [reply data/data/powervs/loadbalancer/versions.tf]: @mkumatag we can't have one global provider. It fails with this
```
│ Error: Failed to query available provider packages
│
│ Could not retrieve the list of available versions for provider hashicorp/ibm: provider registry registry.terraform.io does not have a provider named registry.terraform.io/hashicorp/ibm
│
│ Did you intend to use ibm-cloud/ibm? If so, you must specify that source address in each module which requires that provider. To see which modules are currently depending on
│ hashicorp/ibm, run the following command:
│     terraform providers
```

File: data/data/powervs/variables.tf (hunk 6 above)
Comment (bpradipt) [reply data/data/powervs/variables.tf]: I'll remove the default for now.

File: data/data/powervs/loadbalancer/alb.tf (hunk 2 above)
Comment (bpradipt) [reply data/data/powervs/loadbalancer/alb.tf]: depends on OpenShift allowing different number of masters :-) 
I'll change it derive the count from master.ips.

File: data/data/powervs/loadbalancer/versions.tf (hunk 5 above)
Comment (mkumatag) [reply data/data/powervs/loadbalancer/versions.tf]: Interesting, I'm wondering how we are managing this in the UPI scripts? because this is going to be challenging to maintain the same version everywhere.

File: data/data/powervs/loadbalancer/alb.tf (hunk 4 above)
Comment (bpradipt) [reply data/data/powervs/loadbalancer/alb.tf:7]: @mkumatag is it ok if I address this is different PR? Need to test few scenarios based on the hostname generated and accessibility.

File: data/data/powervs/loadbalancer/alb.tf (hunk 8 above)
Comment (bpradipt) [reply data/data/powervs/loadbalancer/alb.tf]: Done

File: data/data/powervs/loadbalancer/alb.tf (hunk 1 above)
Comment (bpradipt) [reply data/data/powervs/loadbalancer/alb.tf]: Removed it from the files

File: data/data/powervs/loadbalancer/alb.tf (hunk 4 above)
Comment (mkumatag) [reply data/data/powervs/loadbalancer/alb.tf:7]: make sense, will create a separate PR for that.

File: data/data/powervs/loadbalancer/sg.tf [hunk 9]
```diff
@@ -0,0 +1,26 @@
+locals {
+  tcp_ports = [22623, 6443, 443, 80]
+}
+data "ibm_is_vpc" "vpc" {
+  name = var.vpc_name
+}
+
+resource "ibm_is_security_group" "ocp_security_group" {
+  name = "${var.cluster_id}-ocp-sec-group"
+  vpc  = data.ibm_is_vpc.vpc.id
```
Comment (mkumatag) [inline data/data/powervs/loadbalancer/sg.tf:10]: Need tag here as well

Comment (bpradipt) [reply data/data/powervs/loadbalancer/sg.tf:10]: done

**Review by mkumatag — APPROVED** [review]
LGTM

File: data/data/powervs/loadbalancer/versions.tf (hunk 5 above)
Comment (clnperez) [reply data/data/powervs/loadbalancer/versions.tf]: iiuc we won't have to carry any of these provider files forward, and can remove them when we're relying on the installer to set up the tf env.
